sql/core/src/test/scala/org/apache/spark/sql/streaming/EventTimeWatermarkSuite.scala

```diff
@@ -300,6 +300,67 @@ class EventTimeWatermarkSuite extends StreamTest with BeforeAndAfter with Matche
     )
   }
 
+  test("watermark with 2 streams") {
+    val first = MemoryStream[Int]
+
+    val firstAggregation = first.toDF()
+      .withColumn("eventTime", $"value".cast("timestamp"))
+      .withWatermark("eventTime", "10 seconds")
+      .select('value)
+
+    val second = MemoryStream[Int]
+
+    val secondAggregation = second.toDF()
+      .withColumn("eventTime", $"value".cast("timestamp"))
+      .withWatermark("eventTime", "10 seconds")
+      .select('value)
+
+    val union = firstAggregation.union(secondAggregation)
+      .writeStream
+      .format("memory")
+      .queryName("test")
+      .start()
+
+    def populateNewWatermarkFromData(stream: MemoryStream[Int], data: Int*): Unit = {
+      stream.addData(data)
+      union.processAllAvailable()
+      // add a dummy batch so lastExecution has the new watermark
+      stream.addData(0)
+      union.processAllAvailable()
+    }
+
+    def assertQueryWatermark(watermark: Int): Unit = {
+      assert(union.asInstanceOf[StreamingQueryWrapper].streamingQuery
+        .lastExecution.offsetSeqMetadata.batchWatermarkMs
+        == watermark)
+    }
+
+    populateNewWatermarkFromData(first, 11)
```

Comment:
if these are always used together, then these functions can be merged .. right?

Reply:
Also, I think you can use the `testStream..AddData... AssertOnQuery` pattern. its cleaner.
https://github.com/apache/spark/blob/master/sql/core/src/test/scala/org/apache/spark/sql/streaming/StreamingAggregationSuite.scala#L180

Reply:
The problem is that watermark recalculation happens at the beginning of each batch, and to sequence executions I have to call CheckData or CheckLastBatch. So that method ends up producing a test multiple times longer, since a single entry is:

AddData(realData)
CheckLastBatch
AddData(0)
CheckLastBatch
AssertOnQuery

Reply:
even better example - https://github.com/apache/spark/blob/master/sql/core/src/test/scala/org/apache/spark/sql/streaming/EventTimeWatermarkSuite.scala#L107

Reply:
You are right. what you have is better.